ziti/cmd/ascode/ascode_test.go

```diff
@@ -0,0 +1,238 @@
+	limitations under the License.
+*/
+
+package cmd
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/antchfx/jsonquery"
+	"github.com/michaelquigley/pfxlog"
+	"github.com/openziti/ziti/ziti/cmd/ascode/download"
+	"github.com/openziti/ziti/ziti/cmd/ascode/upload"
+	"github.com/openziti/ziti/ziti/cmd/edge"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+)
+
+var log = pfxlog.Logger()
+
+func TestYamlUploadAndDownload(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cmdComplete := make(chan bool)
+	qsCmd := edge.NewQuickStartCmd(os.Stdout, os.Stderr, ctx)
+
+	tmp, _ := os.MkdirTemp("/tmp", "ziti-")
+	qsCmd.SetArgs([]string{"--home", tmp, "--ctrl-address", "127.0.0.1"})
+	os.Setenv("ZITI_CONFIG_DIR", tmp)
+	os.Setenv("ZITI_HOME", tmp)
+
+	go func() {
+		err := qsCmd.Execute()
+		if err != nil {
+			log.Fatal(err)
+		}
+		cmdComplete <- true
+	}()
+
+	c := make(chan struct{})
+	go waitForController("https://127.0.0.1:1280", c)
+	timeout, _ := time.ParseDuration("180000000s")
+	select {
+	case <-c:
+		//completed normally
+		log.Info("controller online")
+	case <-time.After(timeout):
+		cancel()
+		panic("timed out waiting for controller")
+	}
+
+	login := edge.NewLoginCmd(os.Stdout, os.Stderr)
+	login.SetArgs([]string{"127.0.0.1:1280", "-y", "--username", "admin", "--password", "admin"})
+	err := login.Execute()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	performTest(t)
+
+	cancel() //terminate the running ctrl/router
+
+	<-cmdComplete
+	fmt.Println("Operation completed")
+}
+
+func waitForController(ctrlUrl string, done chan struct{}) {
+	tr := &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
+	client := &http.Client{Transport: tr}
+	for {
+		r, e := client.Get(ctrlUrl)
+		if e != nil || r == nil || r.StatusCode != 200 {
+			time.Sleep(50 * time.Millisecond)
+		} else {
+			break
+		}
+	}
+	done <- struct{}{}
+
+}
+
+func performTest(t *testing.T) {
+
+	errWriter := strings.Builder{}
+
+	uploadWriter := strings.Builder{}
+	uploadCmd := upload.NewUploadCmd(&uploadWriter, &errWriter)
+	uploadCmd.SetArgs([]string{"--verbose", "--yaml", "./test.yaml"})
+
+	err := uploadCmd.Execute()
+	if err != nil {
+		t.Fail()
+		log.Fatal(err)
+	}
+
+	downloadWriter := strings.Builder{}
+	downloadCmd := download.NewDownloadCmd(&downloadWriter, &errWriter)
+	downloadCmd.SetArgs([]string{"--json"})
+
+	err = downloadCmd.Execute()
+	if err != nil {
+		t.Fail()
+		log.Fatal(err)
+	}
+
+	result := downloadWriter.String()
+
+	log.Debug("Read " + result)
+
+	doc, err := jsonquery.Parse(strings.NewReader(result))
+	if err != nil {
+		panic(err)
+	}
+
+	assert.NotEqual(t, 0, len(doc.ChildNodes()))
+	if len(doc.ChildNodes()) == 0 {
+		// no data, nothing to test
+		return
+	}
+
+	externalJwtSigner1 := jsonquery.FindOne(doc, "//externalJwtSigners/*[name='NetFoundry Console Integration External JWT Signer']")
```

Comment:
would you add a test case for whatever the issue was with "handling missing identities" from https://github.com/openziti/ziti/pull/2571/commits/498c206d76e1b45d28738cc1f1ce0f33b401052c ?